### PR TITLE
feat(notifications): the write path and the four inbox routes (#1598)

### PR DESCRIPTION
**Pull request by MichalZelasko**
### Summary

Phase 2 of the notification center's implementation, stacked on #1689's schema: the write path and the four inbox routes. Still no producer wired in — that's phase 5 — this is the mechanism every future producer and the frontend inbox will call.

### Business Context

Continues #1637's design-first plan (`docs/design/notification-center-plan.md`) for #1598, following the plan's own phased work breakdown (item 2).

### Changes

- **Decision 1's catalog** (`backend/app/services/notification_catalog.py`): mandatory-ness and the read-time content gate per event type — narrower than the full plan table on purpose, since producer/audience wiring is each later phase's own job.
- **The write path** (`NotificationCenterService.write`, Decisions 2–4): resolves each recipient's in-app/email preference (the three legacy `User` columns for the four events they already govern, the new `notification_preferences` table for everything else), skips the preference lookup entirely for a mandatory event, consults `rate_limit.py` before a mandatory write keyed on `(actor_user_id, event_type)`, and writes through `INSERT ... ON CONFLICT DO NOTHING` rather than relying on the unique constraint to raise. `use_savepoint` wraps the write in `db.begin_nested()` for a caller (a future `AgentRunnerService.finish` wiring) that cannot afford its own transaction poisoned by a failure here.
- **The four inbox routes** (`GET /notifications`, `GET /notifications/unread-count`, `PATCH /notifications/{id}`, `POST /notifications/mark-all-read`) share one gate-aware predicate (Decision 7): `approval_requested` degrades (the row stays visible, its decide-link is withheld from a non-decider) and every other gated event type excludes outright for a reader who no longer holds it — current org admin/app admin for `security_event`, app admin for `configuration_changed`, `runs:view` for the two report events, `resolve_access` against the named collection for ingestion, and current membership against an announcement's own `audience_spec` (all-organizations, a named list, optionally role-narrowed).

### Verification

- 34 integration tests against a real database: dedup, preference resolution, mandatory bypass, the savepoint under a forced FK violation, and all seven gate variants including the malformed-row edge cases.
- 8 API-level tests through the app with the repository stubbed at the database edge.
- A small catalog unit test.
- The four new platform files (`notification_catalog.py`, `notification_center.py`, `repositories/notification.py`, the routes module) are added to `pyproject.toml`'s coverage/ty include lists and held to 100%.
- `ty check`, `ruff` and the full backend suite (unit + integration, 7870 + 862 passed) are clean. `alembic check` still reports zero drift since this phase adds no schema.

### Linked Issues

Part-of #1598

### CODE REVIEW

**Comment by MichalZelasko** [comment]
@codex review

File: backend/app/api/routes/v1/notifications.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    service: NotificationCenterSvc,
+    ctx: Auth,
+    cursor: str | None = Query(None, description="Opaque page token from a previous response"),
+    limit: int = Query(50, ge=1, le=100, description="Max items to return"),
+) -> Any:
+    """The caller's own inbox, newest first, gate-filtered (Decision 7)."""
+    after = decode_cursor(cursor) if cursor else None
+    rows, strip = await service.list_inbox(ctx, after=after, limit=limit)
+    items = [NotificationRead.from_row(row, strip_context_url=strip[row.id]) for row in rows]
+    next_cursor = encode_cursor(rows[-1].created_at, rows[-1].id) if len(rows) == limit else None
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/api/routes/v1/notifications.py]: Confirmed and fixed in `notification_center.py`'s `list_inbox`: it now returns a resume cursor distinct from the visible-rows count, so a scan the round cap cut short (rather than genuinely exhausted) still returns a `next_cursor` instead of silently stranding rows. `GET /notifications` now builds `next_cursor` from that resume cursor rather than `len(rows) == limit`. Regression tests added: `test_a_fully_gated_backlog_exhausts_its_fetch_budget` now asserts a resume cursor comes back, and `test_a_second_page_starts_after_the_cursor` asserts the opposite on genuine exhaustion.

File: backend/app/repositories/notification.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    *,
+    notification_id: uuid.UUID,
+    recipient_id: uuid.UUID,
+    organization_id: uuid.UUID,
+) -> Notification | None:
+    return await db.scalar(
+        select(Notification).where(
+            Notification.id == notification_id,
+            Notification.recipient_user_id == recipient_id,
+            or_(
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/repositories/notification.py]: Confirmed and fixed: `get_own` now filters on `Notification.in_app_visible.is_(True)`, matching `list_inbox_page` and `list_unread`. Regression test added: `test_marking_an_email_only_row_reads_as_missing`.

File: backend/app/services/notification_center.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    """
+    return f"{created_at.isoformat()}|{notification_id}"
+
+
+def decode_cursor(raw: str) -> tuple[datetime, uuid.UUID]:
+    try:
+        created_at_raw, id_raw = raw.split("|", 1)
+        return datetime.fromisoformat(created_at_raw), uuid.UUID(id_raw)
+    except ValueError as exc:
+        raise BadRequestError(message="Invalid pagination cursor", details={"cursor": raw}) from exc
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:99]: Stale - already fixed. Current `decode_cursor` (notification_center.py) raises `BadRequestError` when `created_at.tzinfo is None`, with a comment describing exactly this failure mode.

File: backend/app/services/notification_center.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        afford to be poisoned by a failure here - `AgentRunnerService.finish`
+        is exactly such a context (Decision 2). A failure inside the savepoint
+        rolls back to it and is logged as `notification_write_failed`; the
+        caller's transaction proceeds and commits normally. Without it, a
+        failure propagates to the caller, which is correct for a context built
+        around this write succeeding.
+        """
+        mandatory = is_mandatory(event_type)
+        if mandatory and actor_user_id is not None:
+            decision = await rate_limit.consume(
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:191]: Confirmed and fixed: the mandatory-write rate limiter is no longer skipped when `actor_user_id` is `None` (a real case - `AppAdminAuditLog.actor_user_id` is nullable for a system-triggered audit entry). It's now keyed on `actor_user_id or "system"`, so that case is bounded rather than exempted. Regression test added: `test_a_rate_limited_mandatory_write_with_no_actor_still_writes_nothing`.

File: backend/app/repositories/member.py [hunk 5]
```diff
@@ -254,6 +254,84 @@ async def list_app_admin_emails(
...
+    `role`, when given, narrows to that role - an announcement's audience can
+    be narrowed to owners and admins (Decision 5), and a member demoted out of
+    that role no longer qualifies even though their plain membership survives.
+    """
+    conditions = [
+        OrganizationMember.user_id == user_id,
+        OrganizationMember.organization_id.in_(organization_ids),
+    ]
+    if role is not None:
+        conditions.append(OrganizationMember.role == role)
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/repositories/member.py]: Confirmed and fixed: an "admin" announcement audience now includes owners too, matching the `_ESCALATION_ROLES` convention `security_event`'s `org_admins` audience already uses (an owner outranks an admin). `member_repo.has_membership_in_any`/`has_any_membership` now take `roles: list[str] | None` instead of a single `role`. Regression test added: `test_an_admin_narrowed_announcement_reaches_an_owner_too`.

File: backend/app/services/notification_center.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+            self.db, notification.announcement_id
+        )
+        if announcement is None:
+            return False
+        spec = announcement.audience_spec or {}
+        role = spec.get("role")
+        organizations = spec.get("organizations")
+        if organizations == "all":
+            return await member_repo.has_any_membership(self.db, user_id=ctx.user_id, role=role)
+        org_ids = [uuid.UUID(str(value)) for value in (organizations or [])]
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py]: Confirmed and fixed: malformed organization ids in `audience_spec` are now skipped individually (matching `_collections_visible`'s own pattern for the same class of unenforced JSONB) instead of raising an unguarded `ValueError` that would take the rest of the reader's inbox down with it. Regression test added: `test_a_malformed_organization_id_in_the_audience_is_skipped`.

File: backend/app/services/notification_center.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+# notification is skipped over the limit.
+_MANDATORY_WRITE_LIMIT = rate_limit.Limit(attempts=20, window_seconds=60)
+
+# Bounds on the app-side work the gate-aware read paths do, since Decision 7's
+# recheck cannot be pushed into a plain `COUNT`/`UPDATE` - each candidate row
+# needs its own permission check. Generous enough that an ordinary inbox never
+# notices either bound; a pathological backlog is truncated rather than
+# unbounded.
+_MAX_INBOX_FETCH_ROUNDS = 5
+_UNREAD_CANDIDATE_CAP = 500
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:80]: Confirmed, and intentional - documented in the comment above `_UNREAD_CANDIDATE_CAP` as "truncated rather than unbounded". Still a real gap in effect (a >500-unread backlog under-reports the badge and `mark_all_read` leaves some rows unread), just a deliberate tradeoff rather than an oversight. Leaving this open rather than resolving it myself - it's a design decision to revisit (batch through it in rounds like `list_inbox` does) rather than a same-shape bug fix, and that tradeoff isn't mine to relitigate unilaterally while fixing the rest of this review.

File: backend/app/services/notification_center.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+                after=cursor,
+                limit=limit,
+            )
+            if not batch:
+                break
+            for row in batch:
+                gate = await self._gate(ctx, row)
+                if gate.visible:
+                    visible.append(row)
+                    strip[row.id] = gate.strip_context_url
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py]: Confirmed - `_collections_visible`/`_announcement_visible` do a per-row DB lookup inside `_gate`, called once per candidate in `list_inbox`/`unread_count`/`mark_all_read`'s loops, with no batching or caching. Real, but a bigger lift (batching the gate check means restructuring these three loops to resolve visibility for a whole page/candidate-set at once rather than row-by-row) and it's a performance concern, not a correctness one - leaving this open for a follow-up rather than rushing a restructure into this fix pass.

File: backend/app/services/notification_center.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+        self,
+        *,
+        recipients: list[uuid.UUID],
+        event_type: NotificationEventType,
+        occurrence_id: str,
+        summary: str,
+        context_url: str | None = None,
+        render_context: dict[str, Any] | None = None,
+        organization_id: uuid.UUID | None = None,
+        announcement_id: uuid.UUID | None = None,
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:161]: Confirmed - `write()` takes `organization_id: UUID | None` with no check that it matches what `event_type` expects. Investigated a fix and it's genuinely ambiguous, not just unenforced: `configuration_changed`/`announcement` are always org=None (deployment-wide), but `security_event` is legitimately *both* - org-scoped for an organization's own secret, org=None for an app-admin-audience row like impersonation or user management (see `_security_audience`'s own docstring in `notifications.py`). A correct allow-list would need a per-event-type policy this catalog doesn't have yet (`notification_catalog.py` only has mandatory-ness and the read-time gate), so encoding one under this fix pass risked getting `security_event`'s dual-mode wrong. Leaving this open rather than a quick assertion that would be wrong for one of the two mandatory event types.

File: backend/app/services/notification_center.py [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+        one's `_Gate` - what to withhold or override when serializing it."""
+        user_id = self._require_caller(ctx)
+        visible: list[Notification] = []
+        gates: dict[uuid.UUID, _Gate] = {}
+        cursor = after
+        for _ in range(_MAX_INBOX_FETCH_ROUNDS):
+            batch = await notification_repo.list_inbox_page(
+                self.db,
+                recipient_id=user_id,
+                organization_id=ctx.organization_id,
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:335]: Confirmed - a genuine gap, but not one this PR can safely fix on its own. `_gate`'s `ORG_ADMIN_OR_APP_ADMIN` case correctly says an app admin can see an org-scoped `security_event` row regardless of membership, but `get_active_organization` (`api/deps.py`) 404s any caller - app admin or not - who sets `X-Organization-Id` to an org they don't belong to, so that org can never become `ctx.organization_id` for them and the row is gate-visible but never fetched. Fixing this means changing `get_active_organization` to special-case `is_app_admin`, which is a load-bearing dependency for *every* org-scoped route in the app, not something specific to notifications - widening app-admin access to arbitrary organizations deserves its own deliberate decision (and probably its own narrower mechanism - an admin-only cross-org read path rather than loosening the general org-header dependency) rather than a side effect of a PR review fix here. Leaving this open and flagging it as worth its own issue.

File: backend/app/services/notification_center.py [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+        context_url: str | None,
+        render_context: dict[str, Any] | None,
+        organization_id: uuid.UUID | None,
+        announcement_id: uuid.UUID | None,
+        mandatory: bool,
+    ) -> list[Notification]:
+        written: list[Notification] = []
+        for recipient_id in recipients:
+            in_app_visible = mandatory or await self._channel_enabled(
+                recipient_id, event_type, NotificationChannel.IN_APP
```
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py]: Confirmed - `_write_rows` does one awaited insert pair per recipient with no bulk insert. Real, but the same class of deferred-perf issue as the content-gate batching above (a bigger restructure, not blocking correctness), and this method also backs every event type including the mandatory/rate-limited ones - batching it safely needs some care around partial-failure semantics (what a bulk insert does when one recipient's row would violate the dedup constraint, versus `insert_notification_if_new`'s current per-row `ON CONFLICT` skip). Leaving this open for a follow-up.

File: backend/app/services/notification_center.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py]: Fixed: `_write_rows`'s content-gate now resolves every `COLLECTIONS_VIEW`/`ANNOUNCEMENT_AUDIENCE` row's dependency once per batch (`_build_gate_cache`, one query per *distinct* collection/announcement in the page) rather than once per row. Landed in 0df7f6cf on this branch.

File: backend/app/services/notification_center.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:161]: Fixed: `write()` now raises `AssertionError` when `organization_id is None` for any event type outside `_DEPLOYMENT_WIDE_ELIGIBLE` (`SECURITY_EVENT`, `CONFIGURATION_CHANGED`, `ANNOUNCEMENT`). Landed in 0df7f6cf on this branch.

File: backend/app/services/notification_center.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:335]: Fixed: `list_inbox_page`/`list_unread`/`get_own` now take `is_app_admin: bool` and drop the org-scope predicate entirely for an app admin, rather than narrowing to the active organization - checking `ctx.is_app_admin` explicitly, as suggested. Landed in 0df7f6cf on this branch.

File: backend/app/services/notification_center.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py]: Fixed: `_write_rows` now batches the preference reads before the per-recipient loop - one `get_channel_preferences`/`get_legacy_email_preferences` call per fan-out instead of two per recipient. The insert side stays per-recipient (dedup-on-conflict is what makes the `begin_nested()` savepoint isolation work), so this removes the majority of the N+1 cost without touching the error-isolation model. Landed in 0df7f6cf on this branch.

File: backend/app/services/notification_center.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MichalZelasko) [reply backend/app/services/notification_center.py:80]: Deliberately left as-is for this pass - fixing this properly means either bounded multi-round fetching (extending `list_inbox`'s own `_MAX_INBOX_FETCH_ROUNDS` pattern to `unread_count`/`mark_all_read`) or exposing the cap explicitly in the response shape, and both are more than this stacked change should carry. Documented and tracked as vstorm-co/agenticos#1761 (severity:low), not fixed in this pass.

File: backend/app/services/notification_center.py [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+                gate is ContentGate.ANNOUNCEMENT_AUDIENCE
+                and notification.announcement_id is not None
+            ):
+                announcement_ids.add(notification.announcement_id)
+
+        kb_by_id = await knowledge_base_repo.get_by_ids(self.db, list(collection_ids))
+        kb_access_by_id = {
+            kb_id: await resolve_access(
+                self.db, ctx, kb, Perm.COLLECTIONS_VIEW, resource_type=COLLECTION
+            )
```
[the remark this answers is not in the document]
Comment (DEENUU1) [reply backend/app/services/notification_center.py:462]: Confirmed and fixed in `9a65ab73`.

`_ingestion_audience` falls back to `_administrator_ids` when nobody initiated
the ingestion, and that set includes every app admin - who hold no membership
row anywhere, so `readable_kb`'s `resolve_access` refused the collection on the
organization mismatch alone and hid a row they were deliberately written.

`_collections_visible` now admits an app admin outright, which is the same
admission `ORG_ADMIN_OR_APP_ADMIN`, `APP_ADMIN` and
`notification_delivery._current_role` already give them. Covered by
`test_an_app_admin_sees_the_ingestion_notice_they_were_addressed`.

File: backend/app/services/notification_center.py [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+        # `notifications.py`) - an owner outranks an admin, not a role an
+        # "admin"-only match would silently exclude from their own audience.
+        roles = (
+            list(_ESCALATION_ROLES) if role == OrgRoleName.ADMIN.value else [role] if role else None
+        )
+        organizations = spec.get("organizations")
+        if organizations == "all":
+            return await member_repo.has_any_membership(self.db, user_id=ctx.user_id, roles=roles)
+        org_ids = []
+        for value in organizations or []:
```
[the remark this answers is not in the document]
Comment (DEENUU1) [reply backend/app/services/notification_center.py:553]: Confirmed and fixed in `9a65ab73`. `audience_spec` is unenforced JSONB, and the
element-level `except ValueError` below defended only against a malformed
*entry* - the collection itself can be any JSON value, and iterating a number
raises a `TypeError` that takes down every inbox read touching the row.

`_resolve_announcement_visible` now returns `False` for anything that is not a
list, one level up from the existing guard and for the same reason. Covered by
`test_an_audience_that_is_not_a_collection_at_all_is_skipped`.

File: backend/app/services/notification_center.py [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+            "render_context": render_context,
+            "organization_id": organization_id,
+            "announcement_id": announcement_id,
+            "mandatory": mandatory,
+        }
+        if not use_savepoint:
+            return await self._write_rows(**kwargs)
+        try:
+            async with self.db.begin_nested():
+                return await self._write_rows(**kwargs)
```
[the remark this answers is not in the document]
Comment (DEENUU1) [reply backend/app/services/notification_center.py:218]: Not reproducible against this branch - the savepoint is already nested *per
recipient*, not once around the fan-out.

`_write_rows` opens `async with self.db.begin_nested()` inside the
`for recipient_id in recipients` loop, and its `except Exception` logs
`notification_write_failed` and `continue`s to the next recipient. A deleted
recipient's FK violation rolls back that recipient's savepoint alone;
`test_one_bad_recipient_does_not_discard_the_others` asserts exactly the
scenario described - a good recipient and a missing one in the same call, with
the good one's row surviving.

What *was* outside a savepoint is the preference read that runs before the
loop, reported separately on #1696 and fixed in `9a65ab73`.

**Comment by DEENUU1** [comment]
Merged as part of #1760, which carries all fourteen steps of #1598 in one
squash - see 2369aaef on `main`.

Not a rejection of this step, and not a decision about how the work was split:
GitHub's own "stacked pull requests" feature refused to merge the chain one
pull request at a time, answering
`This pull request is part of a stack and must be merged using the
asynchronous merge REST API` - and that API answers 403 for this repository,
while retargeting anything below the top three was refused outright. The tip
branch already contained every commit of every branch in the stack, so the
whole feature landed from there.

This branch's commits are in `main` with their own history intact; only the
pull request closes unmerged. The review threads opened here were triaged
rather than bulk-resolved: each one has a reply saying what was fixed, or the
evidence for rejecting it.
